Add drawing of Statevector using ket notation

peendebak opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Add drawing of the qiskit `StateVector` as LaTeX with the standard ket notation. A statevector `StateVector([0.7071, -0.7071])` will be drawn as (rendered in jupyter notebook as formatted latex)
```
0.7071 |0> - 0.7071 |1>
```
The old behaviour to plot as a vector is still available using the `convention='vector'` option.

### Details and comments

The qiskit `StateVector` can already be plotted with various output modes, but not yet with a compact notation.
An example of the new style:
![image](https://user-images.githubusercontent.com/17724047/118363378-6b01cf00-b594-11eb-8baa-a8a20315c570.png)

jaygambetta commented:
if using latex I would also use fractions with sqrt 2 etc, phases (e^{I pi}, e^{I pi/2} etc) . Also I don't really see why we need latex and latex_ket can't all these improvements go into latex. I am assuming the point of the latex output is to make an appealing state which could be put into a latex document, otherwise the user should use text.

peendebak commented:
@jaygambetta If preferred I can merge `latex` and `latex_ket`. The notation will then be something like `Statevector.draw(output='latex', convention='ket')`, e.g.

![image](https://user-images.githubusercontent.com/17724047/114281992-c09d0600-9a41-11eb-8622-3302e3fbdedb.png)

Let me know your thoughts on that.

I would not now how to convert the entries of the `Statevector`  (which are floating point numbers) to an exact format (e.g. one involving `sqrt`). I could imagine a `Statevector` with entries in `sympy` format being converted properly, but the `Statevector` class converts input already to a complex float:
![image](https://user-images.githubusercontent.com/17724047/114281933-7ddb2e00-9a41-11eb-9dd2-6ba5bc4bcebc.png)

kdk commented:
Agree with merging this into `output='latex'` . Adding a `convention` kwarg seems reasonable, though for me, the ket notation is nicer and maybe a better default than the array (though maybe @frankharkins can comment, I think this originally came from the textbook).

Checking for sqrt 2 and phases is possible (we have similar code to find multiples of pi https://github.com/Qiskit/qiskit-terra/blob/21b40bf/qiskit/circuit/tools/pi_check.py#L28 ) but not necessarily trivial. This could be left for a separate issue.

eendebakpt commented:
@kdk I merged the `latex` and `latex_ket` options, setting the default convention to `ket` (which is different from the old defaults). With `sympy` I convert the simple cases such as `np.sqrt(2)/2` to nice formatting.

![image](https://user-images.githubusercontent.com/883786/115971244-abe36680-a547-11eb-9b84-ea39209805a8.png)

frankharkins commented:
This is cool and I think probably better default than the arrays for statevectors. The code that converts a number into something more human-readable is [here](https://github.com/Qiskit/qiskit-terra/blob/82635d57a83d522c63044bd5423a38f3186428ce/qiskit/visualization/array.py#L21), and I think this should be modified and used for both functions instead of having it in two places. Also agree the form $e^{i\*pi\*x}$ is probably more useful for most cases.

peendebak commented:
@frankharkins @kdk I converted this PR to RFC for now.

The method `_num_to_latex` from `array.py` performs a task similar to `num_to_latex_ket` in the current version of the PR. The logic in `_num_to_latex` is a bit different from my method though. For a vector we want to translate [0, -1] into "[0, -1]", but for a ket we want -|1>, so we generate "-" instead of "-1". This makes the merging of the two is a bit non-trivial. I created more unit tests, so if we decide to merge that we can more easily catch all the corner cases.

The method `_num_to_latex` has the additional feature of factoring out a common term in case the real and imaginary part have the same absolute value. The only cases that are processed here are multiples of `1+i`, `1-i`, `-1+i` and `-1-i` I think. These are exactly the cases that would be caught if we try to factor out terms of the form `$r e^{i*pi*x}$` (e.g. `x=1/4`). 
But the factoring out of a common term can not readily be combined with the `nsimplify`. For example cases like `(1+sqrt(2))( (-1-sqrt(5)) - i(1+sqrt(5)) )` will require additional brackets to print them correctly. (I admit that it seems unlikely to encounter this example in the context of coefficients of a statevector/density matrix, but I hope this makes the point clear)

peendebak commented:
Rebased on `main`

peendebak commented:
@nkanazawa1989 Could you please review this PR?

kdk commented on releasenotes/notes/draw-statevector-in-ket-notation-0726959d1f6ea3ce.yaml:
```diff
@@ -0,0 +1,7 @@
+    Allow drawing of a Statevector in latex both in vector and ket notation.
+    In ket notation the entries of the statevector are processed such that 
+    exact factors like fractions or sqrt of two are drawn as such.
+    The old notation is still available as an option.
```
Can you add a comment on how to enable the old convention?

kdk commented on qiskit/visualization/state_visualization.py:
```diff
@@ -18,11 +18,14 @@
 from functools import reduce
 import colorsys
 import numpy as np
+import sympy
```
Can this be moved to a runtime import? (sympy can take a fair amount of time to import, so we try to keep it out of the set of dependencies that are pulled in when someone runs `import qiskit`.)

kdk commented on qiskit/visualization/state_visualization.py:
```diff
@@ -1131,10 +1139,105 @@ def state_to_latex(state, dims=None, **args):
         suffix = f"\\\\\n\\text{{dims={dims_str}}}\n\\end{{align}}"
-    latex_str = array_to_latex(state._data, source=True, **args)
+
+    operator_shape = state._op_shape
```
Is it possible to avoid accessing a private property here? Also, can you add a comment explaining the logic below?

peendebak replied:
The private `_op_shape` is also used by other methods in this file. To avoid this, one should create public methods on the base class `QuantumState`. I think this should be done in a separate PR.
I added some comments to explain the logic.

kdk commented on qiskit/visualization/state_visualization.py:
```diff
@@ -1131,10 +1139,105 @@ def state_to_latex(state, dims=None, **args):
+    Returns:
+        String with latex code or None if no term is required
+    """
+    raw_value = complex(np.real_if_close(raw_value))
```
`np.real_if_close` will truncate only the imaginary parts that are close to zero. Is there a reason not to do the same for the real parts?

peendebak replied:
Nice catch. Added the same functionality for the real parts with two more tests.

kdk commented on qiskit/visualization/state_visualization.py:
```diff
@@ -1131,10 +1139,105 @@ def state_to_latex(state, dims=None, **args):
+    if np.abs(value) > 0:
+        latex_element = sympy.latex(value, full_prec=False)
+        two_term = real_value != 0 and imag_value != 0
+        if isinstance(value, sympy.core.Add):
+            two_term = True
```
Is this redundant of the line above? Or is there another case where `value` could be an `Add`? Can you add a comment?

peendebak replied:
Cases like 1+sqrt(2) can happen. I added a comment

kdk reviewed (commented):
This looks great @peendebak ! I had a few questions below, but the overall approach looks good to me.

peendebak commented:
> This looks great @peendebak ! I had a few questions below, but the overall approach looks good to me.

@kdk I addressed the review comments. Thanks for the feedback.

eendebakpt commented:
@kdk I rebased on main and tests are passing again. Let me know whether I need to make additional changes in order for the PR to be approved.

eendebakpt commented:
@chriseclectic @nonhermitian Could either one of you please review the PR?

nonhermitian commented on qiskit/visualization/state_visualization.py:
```diff
@@ -1149,10 +1156,111 @@ def state_to_latex(state, dims=None, **args):
+    Returns:
+        String with latex code or None if no term is required
+    """
+    import sympy  # runtime import
```
I thought we were trying to remove SymPy from everywhere?

nonhermitian commented on qiskit/visualization/state_visualization.py:
```diff
@@ -1149,10 +1156,111 @@ def state_to_latex(state, dims=None, **args):
+    num = int(np.log2(len(data)))
+    ket_names = [bin(i)[2:].zfill(num) for i in range(2 ** num)]
+
+    latex_terms = numbers_to_latex_terms(data)
```
What happens when things get large?  e.g. `2**27` elements?  What should the output look like in this case?

nonhermitian reviewed (commented):
A few comments.  Also what does output look like for a more common example where a statevector is obtained from a circuit (Unitary)?  In cases like that one gets things like `-0.08190622-0.05512126j` that are not pretty,  Does this clean those up? e.g. convert to amplitude and phase.  Saem for small numbers, e.g. `0.000000312` are those converted to scientific notation for brevity?

eendebakpt commented:
@nonhermitian Thanks for the comments. This PR started with the idea to replace the vector notation with ket notation, which is more compact for many state vectors. Based on comments from @kdk and @jaygambetta I used sympy to create much nicer representations of the entries of the kets. The output should be at least as good as the original output, with the option to output in either vector or ket notation. All further improvements I would prefer to do in a separate PR.

- For the cases you mention I included the output of the method below. Large and small numbers are already convered. A value like `-0.08190622-0.05512126j` is not converted to amplitude/phase, but doing so consistently would be hard to do consistently (we don't want to convert `-1 |0>` to `e^{pi i} |0>` I would say). 

![image](https://user-images.githubusercontent.com/883786/122750820-484b9000-d28f-11eb-9488-6ed6694dee69.png)

- I was not aware of a goal to remove usage of sympy. Based on comments I did make runtime imports.

nonhermitian commented:
Ok answered some questions.  The 2^27 was asked because there is no option that I see to truncate output if the vector is large.

eendebakpt commented:
> Ok answered some questions. The 2^27 was asked because there is no option that I see to truncate output if the vector is large.

@nonhermitian Ah I see the point now. For sparse vectors the ket notation automatically is truncated. I included an optimization to make this plotting more efficient. 

![image](https://user-images.githubusercontent.com/883786/122763240-64a2f900-d29e-11eb-8bd9-252f465e7221.png)

The non-sparse case (e.g. a uniform superposition of stages) for a large statevector is currently fully rendered.  For the old notation the truncation is automatic (the user cannot specify options here). I would be fine with leaving it that way and letting the user specify the other notation for these cases.

If we do want something of truncation, then what about the following approach:

1. Calculate the number of non-zero terms. 
2. If this number is larger than 32, then only plot the first 3 and the final term, with ellipsis in between.

nonhermitian commented:
Yes, some kind of truncation needs to be done.  Similar to the vector notation seems to make the most sense given you want uniformity.

peendebak commented:
@nonhermitian I added truncation to the ket notation. I also added some optimizations such that state vectors of large size (e.g. 2**17) are plotted in a reasonable amount of time.

peendebak commented:
@nonhermitian I rebased to main again. Could you please review the PR once more?

nonhermitian reviewed (approved):
Now that things get truncated properly, I am good with this.  

I had no idea that the `StateVector` class had so many draw options, and wonder what the use cases are of those.  However, it also does not really harm anything to have the options there.

peendebak commented:
@kdk Could you have a look at the PR once more? Thanks

kdk reviewed (approved):
Thanks for the updates here @peendebak , this looks good.
